internals/overlord/state/notices.go

```diff
@@ -0,0 +1,423 @@
+	}
+
+	s.writing()
+
+	now := options.Time
+	if now.IsZero() {
+		now = time.Now()
+	}
+	now = now.UTC()
+	newOrRepeated := false
+	uniqueKey := noticeKey{noticeType, key}
+	notice, ok := s.notices[uniqueKey]
+	if !ok {
+		// First occurrence of this notice type+key
+		s.lastNoticeId++
+		notice = &Notice{
+			id:            strconv.Itoa(s.lastNoticeId),
+			noticeType:    noticeType,
+			key:           key,
+			firstOccurred: now,
+			lastRepeated:  now,
+			expireAfter:   defaultNoticeExpireAfter,
+			occurrences:   1,
+		}
+		s.notices[uniqueKey] = notice
+		newOrRepeated = true
+	} else {
+		// Additional occurrence, update existing notice
+		notice.occurrences++
+		if options.RepeatAfter == 0 || now.After(notice.lastRepeated.Add(options.RepeatAfter)) {
+			// Update last repeated time if repeat-after time has elapsed (or is zero)
+			notice.lastRepeated = now
+			newOrRepeated = true
+		}
+	}
+	notice.lastOccurred = now
+	notice.lastData = options.Data
+	notice.repeatAfter = options.RepeatAfter
+
+	if newOrRepeated {
+		s.noticeCond.Broadcast()
+	}
+
+	return notice.id, nil
+}
+
+func validateNotice(noticeType NoticeType, key string, options *AddNoticeOptions) error {
+	if !noticeType.Valid() {
+		return fmt.Errorf("internal error: attempted to add notice with invalid type %q", noticeType)
+	}
+	if key == "" {
+		return fmt.Errorf("internal error: attempted to add %s notice with invalid key %q", noticeType, key)
+	}
+	return nil
+}
+
+type noticeKey struct {
+	noticeType NoticeType
+	key        string
+}
+
+// NoticeFilter allows filtering notices by various fields.
+type NoticeFilter struct {
+	// Types, if not empty, includes only notices whose type is one of these.
+	Types []NoticeType
+
+	// Keys, if not empty, includes only notices whose key is one of these.
+	Keys []string
+
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// matches reports whether the notice n matches this filter
+func (f *NoticeFilter) matches(n *Notice) bool {
+	if f == nil {
+		return true
+	}
+	// Can't use strutil.ListContains as Types is []NoticeType, not []string
+	if len(f.Types) > 0 && !sliceContains(f.Types, n.noticeType) {
+		return false
+	}
+	if len(f.Keys) > 0 && !sliceContains(f.Keys, n.key) {
+		return false
+	}
+	if !f.After.IsZero() && !n.lastRepeated.After(f.After) {
+		return false
+	}
+	return true
+}
+
+func sliceContains[T comparable](haystack []T, needle T) bool {
+	for _, v := range haystack {
+		if v == needle {
+			return true
+		}
+	}
+	return false
+}
+
+// Notices returns the list of notices that match the filter (if any),
+// ordered by the last-repeated time.
+func (s *State) Notices(filter *NoticeFilter) []*Notice {
+	s.reading()
+
+	notices := s.flattenNotices(filter)
+	sort.Slice(notices, func(i, j int) bool {
+		return notices[i].lastRepeated.Before(notices[j].lastRepeated)
+	})
+	return notices
+}
+
+// Notice return a single notice by ID, or nil if not found.
```

Comment:
s/return/returns/

Reply:
Good catch -- fixed as a drive-by in the [API PR](https://github.com/canonical/pebble/pull/296/commits/cec77391e82b54118a88db42438d7b43af6c54f5).